### PR TITLE
Add hook for CodegenDecorators to insert a custom symbol provider

### PR DESCRIPTION
**Pull request by mcmasn-amzn**
## Motivation and Context
https://github.com/awslabs/smithy-rs/issues/2158

## Description
This provides a new hook on the `CoreCodegenDecorator` interface. This allows authors of custom decorators (like myself) to manipulate symbols when they are being built.

For example -- this allows me to insert extra derives (such as `serde::Deserialize`).

## Testing
* Ensured all tests are passing locally `./gradlew test`
* Rebuilt the package with my custom decorator and verified my decorator receives the expected callback.

## Checklist
No updates to changelog. This is an internal change that shouldn't affect most users.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/customize/CoreCodegenDecorator.kt [hunk 1]
```diff
@@ -150,6 +156,11 @@ abstract class CombinedCoreCodegenDecorator<CodegenContext, Decorator : CoreCode
     final override fun extraSections(codegenContext: CodegenContext): List<AdHocCustomization> =
         addCustomizations { decorator -> decorator.extraSections(codegenContext) }
 
+    final override fun extraSymbolProvider(base: RustSymbolProvider, model: Model): RustSymbolProvider =
```
[the remark this answers is not in the document]
Comment (mcmasn-amzn) [reply codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/customize/CoreCodegenDecorator.kt]: Sure, no problem. I pushed an update to drop the 'extra' prefix.

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/customize/CoreCodegenDecorator.kt [hunk 2]
```diff
@@ -94,6 +95,11 @@ interface CoreCodegenDecorator<CodegenContext> {
      * Extra sections allow one decorator to influence another. This is intended to be used by querying the `rootDecorator`
      */
     fun extraSections(codegenContext: CodegenContext): List<AdHocCustomization> = listOf()
+
+    /**
+     * Hook for customizing symbols by inserting an additional symbol provider.
+     */
+    fun symbolProvider(base: RustSymbolProvider, model: Model): RustSymbolProvider = base
```
Comment (rcoh) [inline codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/customize/CoreCodegenDecorator.kt:102]: I don't think you need `model` here—model is already available on the inner provider

**Review by rcoh — APPROVED** [review]
LGTM—see one suggestion about removing model. Final approval delegated to @jdisanti

**Review by jdisanti — APPROVED** [review]

**Comment by jdisanti** [comment]
> I see all CI checks pass except 1, "CI (from fork) / ci / Check SDK Examples (pull_request) ". I don't understand the failure on this check. Is this related to my change? Or is it a known issue in this test?

This failure isn't related to your PR, and won't block merging.

**Comment by mcmasn-amzn** [comment]
cool, thank you for the review @jdisanti

**Pull request merged**